[TOAZ-15][TUIM-1] Swap out oauth library to eventually support B2C

rtitle opened the pull request:
[Design doc](https://docs.google.com/document/d/1fnfLSbHWMKi8F__zJm-p7C2URlXIEKzB0-bgyrf-qBg/edit#)
[Ticket](https://broadworkbench.atlassian.net/browse/TOAZ-15)

This PR does the following things:
* Use [react-oidc-context](https://github.com/authts/react-oidc-context) for oauth instead of google-api-client.
* Update `auth.js` to make use of the new OAuth library. We use it in a similar way as before: install listeners on auth events and update state atoms. The nice thing about this is by keeping the same state information most components don't need to change at all.
* Update `SignInButton.js` -- instead of using the Google button we make our own button which opens the sign in pop-up when clicked.

This PR does NOT switch Terra UI to B2C auth -- we are still using Google auth, just with a different library. We'll likely enable B2C in dev only in a separate PR (once a few rough edges are fixed).

Things I've tried to test:
1. Sign in, sign out, all screens work as intended. Integration tests passed!
2. `#billing` page elevates the scope if needed
3. Google token gets auto-renewed each hour
4. `IdleStatusMonitor` logs out the user if enabled

<!--
Hello, friend!
Remember to mention what sort of testing/verification you performed in the course of building this PR, i.e. checking functionality in the browser locally.
Also, so if a screen recording [1] and/or screenshots would be a helpful way to communicate context, please consider including some in your PR description.
Thanks!

[1] https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac
--!>

rtitle committed:
First pass: swap google auth library with react-oidc-context. Working for Google and B2C logins."

rtitle committed:
Clean up and implement scopes

rtitle committed:
Add todo

rtitle committed:
Cleanup and fixes

rtitle committed:
Remove todo

rtitle committed:
Fix up dependencies

rtitle commented on src/components/AuthContainer.js:
```diff
@@ -14,8 +15,11 @@ import TermsOfService from 'src/pages/TermsOfService'
+  const auth = useAuth()
   const authspinner = () => h(centeredSpinner, { style: { position: 'fixed' } })
 
+  useOnMount(() => authStore.update(_.set(['authContext'], auth)))
```
This is grabbing the `AuthContext` out of React context and setting it in our state atom. I do this so I can access it in `src/libs/auth.js` outside of a react component.

Alternatively: we could update every component to call `useAuth()` and pass the `AuthContext` around as needed. That seems to be more how the library is intended to be used; that change would touch a lot more files though. 

See usage examples in https://github.com/authts/react-oidc-context

dmohs replied:
Good call on limiting the scope. We can iterate toward the right approach.

cahrens replied:
Can this be removed now that you have `AuthStoreSetter`?

rtitle replied:
Oops yeah, removed

rtitle commented on src/components/SignInButton.js:
```diff
@@ -1,44 +1,21 @@
-    style: { outlineOffset: 5 }
-  })
+    onClick: () => signIn(false),
+    style: { marginTop: '0.875rem', width: '6rem', height: '2.8rem' }
```
Trying to match the sign in page style here: https://miro.com/app/board/uXjVOZMbcqQ=/

But any styling advice appreciated.

dmohs replied:
I might make it a bit narrower/longer to emphasize, "I am a button."

rtitle commented on src/components/icons.js:
```diff
@@ -65,6 +65,7 @@ export const profilePic = ({ size, style, ...props } = {}) => img({
   src: getUser().imageUrl,
   height: size, width: size,
   style: { borderRadius: '100%', ...style },
+  referrerPolicy: 'no-referrer',
```
I did this to get the profile pic to show up in local testing.

cahrens replied:
I've skimmed through some discussions of this and I _think_ this is fine, but hopefully someone with more understanding of the referrer policy can also double-check.

nmalfroy replied:
I did the same thing in TDR.  This was actually an issue even before the B2C stuff

rtitle commented on src/libs/auth.js:
```diff
@@ -110,24 +131,26 @@ export const initializeAuth = _.memoize(async () => {
   }
-  processUser(getAuthInstance().currentUser.get())
-  getAuthInstance().currentUser.listen(processUser)
+  const userManager = new UserManager(getOidcConfig())
```
Instantiate a `UserManager` directly to bootstrap the initial auth -- but install the listeners on the auth instance from `authStore`.

cahrens replied:
Why is the `OidcConfig` needed both here and in the `Main.js` call to `AuthProvider`?

rtitle replied:
The one in `Main.js` is used for instantiating the `AuthContext`, which is stored in react context and used for all auth operations (sign in/out, listeners, etc).

The one here is directly instantiating a `UserManager` solely to grab the current logged in user (if any) and populate the initial auth state.

For some reason when I _didn't_ do the latter and just used `getAuthInstance()` here instead, the app wouldn't remember my login state. So a page refresh, opening a new tab would require a login. But by instantiating a `UserManager` it remembers my login as expected. I admit I don't fully understand why.

rtitle commented on src/pages/Main.js:
```diff
@@ -14,29 +15,32 @@ import NpsSurvey from 'src/components/NpsSurvey'
-    h(SupportRequest),
-    h(NpsSurvey),
-    h(ConfigOverridesWarning)
+  return h(AuthProvider, getOidcConfig(), [
```
Need to wrap the application in `AuthProvider`. See: https://github.com/authts/react-oidc-context

dmohs replied:
I was wondering if this possibly delayed rendering of the page, but that [does not appear to be the case](https://github.com/authts/react-oidc-context/blob/main/src/AuthProvider.tsx#L223).

rtitle commented on src/pages/SignIn.js:
```diff
@@ -22,11 +20,6 @@ const styles = {
   return h(HeroWrapper, { showMenu: false }, [
     div({ style: { maxWidth: 600 } }, [
-      div({ style: { fontSize: 36, color: colors.dark(0.6) } }, ['New User?']),
-      div({ style: { fontSize: 36, marginBottom: '2rem' } }, [`${getAppName()} requires a Google Account.`]),
```
Removed this text per https://miro.com/app/board/uXjVOZMbcqQ=/

(Soon it won't be true)

petesantos replied:
I would prefer to leave this text in here, especially since we are no longer using the Google sign in button. It may not be clear to a new user that this is currently the only way to sign in. Once we have multiple auth providers we can update the language.

rtitle replied:
Added it back

kyuksel commented:
@rtitle Thank you for including a nice description and annotations. I interpret the PR description as there should be no user-visible changes. Is this correct?

rtitle commented:
@kyuksel there is a user visible change: the `SignInButton` is updated, and some text is removed from the welcome page.

New: 
![image](https://user-images.githubusercontent.com/5368863/165144040-5f4c9a84-29eb-4d30-b587-e7b0d98ccf98.png)

Old:
<img width="791" alt="image" src="https://user-images.githubusercontent.com/5368863/165144093-810bb890-1e22-4fd0-88dd-fe646221ab0b.png">

Aside from that though there are no functional changes: Google sign-in should work the same as before, and all UI screens should still work.

cahrens commented on package.json:
```diff
@@ -28,6 +28,7 @@
     "jszip": "^3.7.1",
     "lodash": "^4.17.21",
     "marked": "^4.0.10",
+    "oidc-client-ts": "^2.0.3",
```
Is there anything to remove with regard to the old auth library? Do we still need to load `https://apis.google.com/js/platform.js` in index.html?

rtitle replied:
Oh, didn't realize it was loaded in index.html. I _think_ it can be removed, but will test.

rtitle replied:
I removed it from index.html.

cahrens commented:
I think we need to touch base with UX about how this should look. Is there someone you have been talking to? If not, I would suggest Lou or Ean.

![image](https://user-images.githubusercontent.com/484484/165311946-c38756e0-973e-421d-a1d3-c6c2c64c154f.png)

cahrens commented:
I think we need to touch base with UX about how this should look. Is there someone you have been talking to? If not, I would suggest Lou or Ean.

![image](https://user-images.githubusercontent.com/484484/165311946-c38756e0-973e-421d-a1d3-c6c2c64c154f.png)

rtitle commented:
> I think we need to touch base with UX about how this should look. Is there someone you have been talking to? If not, I would suggest Lou or Ean.

Good call - I've been working with Lou, will reach out (agree that menu looks off).

rtitle commented:
> When I sign in, the terra UI loads briefly in the frame where I select my account. This seems surprising, and I'm wondering if there is a way we can prevent it.

I noticed that too and am a bit baffled. Wonder if @nmalfroy has any insight..

cahrens commented:
> > When I sign in, the terra UI loads briefly in the frame where I select my account. This seems surprising, and I'm wondering if there is a way we can prevent it.
> 
> I noticed that too and am a bit baffled. Wonder if @nmalfroy has any insight..

Maybe try `signinPopupCallback`?

cahrens commented on src/libs/auth.js:
```diff
@@ -81,14 +104,13 @@ export const bucketBrowserUrl = id => {
-      const profile = user.getBasicProfile()
-      const isSignedIn = user.isSignedIn()
-      //The following few lines of code are to handle sign-in failures due to privacy tools.
+      const isSignedIn = user !== null && user !== undefined
```
Nit: there is an `_.isNil` method you can use for this.

rtitle replied:
Done

cahrens commented on src/libs/auth.js:
```diff
@@ -110,24 +131,26 @@ export const initializeAuth = _.memoize(async () => {
         user: {
-          token: authResponse ? authResponse.access_token : undefined,
-          id: user.getId() || undefined,
+          token: user?.access_token || undefined,
```
If `user` is undefined, `user?.access_token` will return `undefined`. Therefore you don't need the or.

rtitle replied:
Fixed

cahrens commented on src/libs/auth.js:
```diff
@@ -110,24 +131,26 @@ export const initializeAuth = _.memoize(async () => {
-          token: authResponse ? authResponse.access_token : undefined,
-          id: user.getId() || undefined,
+          token: user?.access_token || undefined,
+          id: userId || undefined,
```
I don't think the or is needed here either.

rtitle replied:
Fixed

nmalfroy commented:
> 

Right...what I ended up doing is using a `redirect-from-oauth` and adding a route to show a loader for that redirect.  Not merged yet (working on some UI testing fun) but you can check it out at https://jade-nm.datarepo-dev.broadinstitute.org/

cahrens commented on src/libs/auth.js:
```diff
@@ -17,24 +18,47 @@ import { asyncImportJobStore, authStore, cookieReadyStore, requesterPaysProjectS
+}
+
+export const signIn = includeBillingScope => {
+  if (includeBillingScope) {
```
I'm having trouble with the billing scope. Repro steps:

1) Log in to Terra as Hermione, go to Billing page
2) Pick a project Hermione owns. Note the "No Access" by "Billing Account"-- Hermione should have access.

Looks like:
![image](https://user-images.githubusercontent.com/484484/165326352-e0fdae9f-7d47-42d7-9577-df0d2019a561.png)

Should look like:
![image](https://user-images.githubusercontent.com/484484/165326548-b8c1eb4c-d315-435f-a969-2a1fa9f7d77e.png)

3) Try clicking the edit icon. You get prompted to choose your account again, but it still doesn't correct the issue.

rtitle replied:
Damn I thought this worked.. thanks will look.

rtitle replied:
think I fixed this, it wasn't remembering the scope correctly.

nmalfroy commented:
Lol. I modeled after Lou's mockups.  So you're saying it should be bigger?
:)

On Tue, Apr 26, 2022 at 10:36 AM Christina Ahrens Roberts <
***@***.***> wrote:

> Right...what I ended up doing is using a path /redirect-from-oauth for
> the redirection (though for google I left it as origin.location so I
> wouldn't need to update all of the oauth clients in google) and adding a
> route to show a loader for that redirect. Not merged yet (working on some
> UI testing fun) but you can check it out at
> https://jade-nm.datarepo-dev.broadinstitute.org/
>
> Wow, can we make that Login button larger Nick? :)
>
> [image: image]
> <https://user-images.githubusercontent.com/484484/165324832-6f36f132-6f29-46ed-a6f7-742463c64539.png>
>
> We probably should have consistency between how this page looks and the
> similar Terra page-- can you include that in your discussion with Lou,
> @rtitle <https://github.com/rtitle>?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DataBiosphere/terra-ui/pull/2949#issuecomment-1109876484>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABK7N66ZIO63HHIVUKKOLI3VG75NNANCNFSM5TH52HKQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cahrens commented:
> Lol. I modeled after Lou's mockups. So you're saying it should be bigger? :)
> […](#)

I struck me as rather bloated, but I will defer to UX.

dmohs commented on src/libs/auth.js:
```diff
@@ -17,24 +18,47 @@ import { asyncImportJobStore, authStore, cookieReadyStore, requesterPaysProjectS
+
+export const signIn = includeBillingScope => {
+  if (includeBillingScope) {
+    return getAuthInstance().signinPopup({ scope: 'openid email profile https://www.googleapis.com/auth/cloud-billing' })
```
If you wanted to make each scope a bit more obvious, for example, you could:
```suggestion
    return getAuthInstance().signinPopup({ scope: `
      openid
      email
      profile
      https://www.googleapis.com/auth/cloud-billing
    `.trim().split(/\s+/).join(' ') })
```

rtitle replied:
Done

rtitle commented:
**_enhance_**

dmohs commented on src/libs/auth.js:
```diff
@@ -81,14 +104,13 @@ export const bucketBrowserUrl = id => {
-      const profile = user.getBasicProfile()
-      const isSignedIn = user.isSignedIn()
-      //The following few lines of code are to handle sign-in failures due to privacy tools.
+      const isSignedIn = user !== null && user !== undefined
```
I might:
```suggestion
      const isSignedIn = !_.isNil(user)
```
just to aid readability here.

dmohs reviewed (approved):
This is great. PR notes are helpful. Testing notes were very helpful.

kyuksel commented:
@rtitle Since these changes are overarching, I wonder if we'd gain more confidence about not introducing flakiness by running [test-flakes](https://github.com/DataBiosphere/terra-ui/tree/dev/integration-tests#detecting-flakiness) with at least a few different tests (the default settings for the tool's input parameters should be good). If any of those tests has a pass rate < 100%, it'd be good to run `test-flakes` for that test against `dev` and compare the results. What do you think?

rtitle commented:
> test-flakes
Cool that's a handy utility. Sure I can run that and update with results -- I want to make some of the code changes listed about first though.

cahrens commented:
> > test-flakes
> 
> Cool that's a handy utility. Sure I can run that and update with results -- I want to address some of the feedback above first though.

AFAIK, the integration tests don't ever actually log in using the oath code… they call a shortcut method to fake the login. Therefore I really don't think that anything could be flaky about this.

cahrens commented on src/libs/auth.js:
```diff
@@ -110,24 +131,26 @@ export const initializeAuth = _.memoize(async () => {
+  getAuthInstance().events.addUserUnloaded(processUser)
 })
 
 // This is intended for tests to short circuit the login flow
```
I do wonder if the code down below at line 160 should change.

rtitle replied:
I was thinking about that. Looks like it is just introspecting a provided Google token and updating state, short-circuiting the oauth flow. So it still works today given we still use Google auth. Once we fully switch to B2C we might need a different approach.

petesantos commented on src/pages/Main.js:
```diff
@@ -14,29 +15,32 @@ import NpsSurvey from 'src/components/NpsSurvey'
+        h(TitleManager),
+        h(FirecloudNotification),
+        h(AuthenticatedCookieSetter),
+        h(AuthContainer, [h(Router)])
```
Since we are not using the context, and don't want others to do so, another option could be to do something like:
```javascript      
       h(AuthProvider, getOidcConfig(), [h(AuthStoreSetter)]),
       h(AuthContainer, [h(Router)])
```

rtitle replied:
I like this! Good idea

rtitle replied:
I did this

petesantos commented on src/components/AuthContainer.js:
```diff
@@ -14,8 +15,11 @@ import TermsOfService from 'src/pages/TermsOfService'
   const authspinner = () => h(centeredSpinner, { style: { position: 'fixed' } })
 
+  useOnMount(() => authStore.update(_.set(['authContext'], auth)))
+
```
A companion `AuthStoreSetter` component could be used alongside the suggestion below.
```javascript
export const AuthStoreSetter = () => {
  const auth = useAuth()
  useOnMount(() => authStore.update(_.set(['authContext'], auth)))
  return null
}
```

petesantos reviewed (commented):
I have not full tested this with scopes - but it looks like Christina has done so.

kyuksel commented:
> AFAIK, the integration tests don't ever actually log in using the oath code… they call a shortcut method to fake the login. Therefore I really don't think that anything could be flaky about this.

If that's the case, never mind my previous comment. I didn't get to go through the code changes yet and was just thinking high-level :)

rtitle committed:
Post rebase fixes

rtitle committed:
First round of PR feedback

rtitle committed:
Fix billing scope escalation

rtitle committed:
Break up scopes to new lines for readability

rtitle commented:
I _think_ I've resolved all PR feedback so far, and resolved conversations above. The scope escalation should be fixed, and it now shows a spinner instead of Terra UI flickering in the popup.

I haven't addressed the button styling yet -- I have a pending question out to Lou. Once I'm sure that's settled I'll re-request review.

cahrens commented:
@rtitle The billing project scope escalation still doesn't seem to be behaving properly. On dev, if you log in with Hermione you can see the billing account info immediately. On your branch, when you log in with Hermione it initially says "No Access". When I choose "Change Billing Account" it brings up the Google account selector, and when I select Hermione again the "Change Billing Account" dialog opens, but with no accounts populated. I then have to cancel the dialog and refresh the billing page to get the account information to populate.

![image](https://user-images.githubusercontent.com/484484/166511424-1f305b74-ea94-4ee6-a96c-7a49811fa69e.png)

rtitle committed:
Remove unnecessary authStore udpate

rtitle commented:
Thanks @cahrens. I see the same billing account behavior as you -- looking into it more. I think we probably need the additional sign-in to elevate scopes (at least until we migrate to B2C which can do this internally I think). But the menus should be populated immediately after sign-in without a refresh. I'll do another pass on this tomorrow.

rtitle committed:
Fix billing pages; styling per discussion with UX

rtitle committed:
Fix automatic silent auth renewal

rtitle committed:
Remuse unused parameter

rtitle committed:
Fix silent login

rtitle committed:
Technically only need to wrap AuthStoreSetter in AuthProvider

rtitle commented:
@cahrens (cc @petesantos @dmohs @kyuksel) I think this is ready for another look when you have a chance. Since last time:
* Rebased with dev and resolved conflicts.
* Connected with Lou and updated styles for the sign-in page + button based on [invision screens](https://dspuxteam.invisionapp.com/overview/Azure-login-screens-cl2qcmlvb002c010icf7fez7z/screens). Feel free to double check my CSS choices 😅 
* Fixed billing issues -- I think things populate correctly now when you elevate billing scopes.
* Fixed a bug with automatic login refresh -- I think it works now, I left this branch and dev logged in overnight in separate incognito windows, and they both stayed logged in.
* Various code style improvements from prior feedback.

rtitle committed:
Another token refresh fix

cahrens commented:
@rtitle I'm seeing a couple of console warnings that would be good to investigate. The second is coming from the SignInButton.

````
react_devtools_backend.js:3973 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at TopBar (http://localhost:3000/static/js/src_appLoader_js-data_image_svg_xml_3Csvg_xmlns_27http_www_w3_org_2000_svg_27_viewBox_270_0_1-d7b3bc.chunk.js:13384:30)
    at div
    at div
    at FooterWrapper (http://localhost:3000/static/js/src_appLoader_js-data_image_svg_xml_3Csvg_xmlns_27http_www_w3_org_2000_svg_27_viewBox_270_0_1-d7b3bc.chunk.js:5761:23)
    at HeroWrapper (http://localhost:3000/static/js/src_appLoader_js-data_image_svg_xml_3Csvg_xmlns_27http_www_w3_org_2000_svg_27_viewBox_270_0_1-d7b3bc.chunk.js:16258:32)
    at SignIn
overrideMethod @ react_devtools_backend.js:3973
printWarning @ react-dom.development.js:67
error @ react-dom.development.js:43
warnAboutUpdateOnUnmountedFiberInDEV @ react-dom.development.js:23914
scheduleUpdateOnFiber @ react-dom.development.js:21840
dispatchAction @ react-dom.development.js:16139
(anonymous) @ react-utils.js:150
(anonymous) @ utils.js:21
r @ lodash.min.js:9
_f @ lodash.min.js:84
l @ lodash.min.js:58
next @ utils.js:21
set @ utils.js:38
update @ utils.js:40
processUser @ auth.js:123
raise @ Event.ts:36
load @ UserManagerEvents.ts:54
_callee54$ @ UserManager.ts:415
tryCatch @ runtime.js:63
invoke @ runtime.js:294
(anonymous) @ runtime.js:119
asyncGeneratorStep @ asyncToGenerator.js:3
_next @ asyncToGenerator.js:25
Promise.then (async)
asyncGeneratorStep @ asyncToGenerator.js:13
_next @ asyncToGenerator.js:25
Promise.then (async)
asyncGeneratorStep @ asyncToGenerator.js:13
_next @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
_signinEnd @ UserManager.ts:69
_callee52$ @ UserManager.ts:374
tryCatch @ runtime.js:63
invoke @ runtime.js:294
(anonymous) @ runtime.js:119
asyncGeneratorStep @ asyncToGenerator.js:3
_next @ asyncToGenerator.js:25
Promise.then (async)
asyncGeneratorStep @ asyncToGenerator.js:13
_next @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
_signin @ UserManager.ts:69
_callee44$ @ UserManager.ts:172
tryCatch @ runtime.js:63
invoke @ runtime.js:294
(anonymous) @ runtime.js:119
asyncGeneratorStep @ asyncToGenerator.js:3
_next @ asyncToGenerator.js:25
Promise.then (async)
asyncGeneratorStep @ asyncToGenerator.js:13
_next @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
signinPopup @ UserManager.ts:69
_callee$ @ AuthProvider.tsx:136
tryCatch @ runtime.js:63
invoke @ runtime.js:294
(anonymous) @ runtime.js:119
asyncGeneratorStep @ asyncToGenerator.js:3
_next @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
signIn @ auth.js:64
onClick @ common.js:61
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
invokeGuardedCallbackAndCatchFirstError @ react-dom.development.js:4070
executeDispatch @ react-dom.development.js:8243
processDispatchQueueItemsInOrder @ react-dom.development.js:8275
processDispatchQueue @ react-dom.development.js:8288
dispatchEventsForPlugins @ react-dom.development.js:8299
(anonymous) @ react-dom.development.js:8508
batchedEventUpdates$1 @ react-dom.development.js:22396
batchedEventUpdates @ react-dom.development.js:3745
dispatchEventForPluginEventSystem @ react-dom.development.js:8507
attemptToDispatchEvent @ react-dom.development.js:6005
dispatchEvent @ react-dom.development.js:5924
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
discreteUpdates$1 @ react-dom.development.js:22413
discreteUpdates @ react-dom.development.js:3756
dispatchDiscreteEvent @ react-dom.development.js:5889
Show 34 more frames
react_devtools_backend.js:3973 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at SignInButton (http://localhost:3000/static/js/src_appLoader_js-data_image_svg_xml_3Csvg_xmlns_27http_www_w3_org_2000_svg_27_viewBox_270_0_1-d7b3bc.chunk.js:12001:76)
    at div
    at div
    at div
    at div
    at FooterWrapper (http://localhost:3000/static/js/src_appLoader_js-data_image_svg_xml_3Csvg_xmlns_27http_www_w3_org_2000_svg_27_viewBox_270_0_1-d7b3bc.chunk.js:5761:23)
    at HeroWrapper (http://localhost:3000/static/js/src_appLoader_js-data_image_svg_xml_3Csvg_xmlns_27http_www_w3_org_2000_svg_27_viewBox_270_0_1-d7b3bc.chunk.js:16258:32)
    at SignIn

cahrens commented on src/components/common.js:
```diff
@@ -523,15 +523,27 @@ export const HeroWrapper = ({ showMenu = true, bigSubhead = false, children }) =
-      h1({ style: { fontSize: 54 } }, `Welcome to ${getAppName()}`),
+      Utils.cond(
+        [isTerra(), () => div({ style: { lineHeight: '2rem' } }, [
+          h1({ style: { fontSize: 54 } }, 'Welcome to'),
```
It doesn't make sense to split this into 2`h1` headers-- the text is logically all part of the same page header.

rtitle replied:
I was trying to match the wireframes which has `Terra Community Workbench` on a new line (it doesn't look right with it all on 1 line):

<img width="1279" alt="image" src="https://user-images.githubusercontent.com/5368863/167465606-38bda6db-939c-48a2-9ce6-8e00a1921fe6.png">

Is there a better way to do this?

cahrens replied:
Semantically, the text needs to be within the same `h1`. I would vote that you don't try to control where it wraps, but if you must, put in a line break tag.

cahrens commented on src/components/common.js:
```diff
@@ -523,15 +523,27 @@ export const HeroWrapper = ({ showMenu = true, bigSubhead = false, children }) =
+        () => h1({ style: { fontSize: 54 } }, `Welcome to ${getAppName()}`)),
       div({ style: { margin: '1rem 0', width: 575, ...(bigSubhead ? { fontSize: 20, lineHeight: '28px' } : { fontSize: 16, lineHeight: 1.5 }) } }, [
-        `${getAppName(true)} is a ${Utils.cond(
+        `${getAppName(!isTerra())} is a ${Utils.cond(
```
`getAppName`'s argument is whether you want the long name. I don't understand why we want to show short name for Terra but long name for others.

rtitle replied:
Again was trying to match the wireframes which has the long name in the title (`Terra Community Workbench`) and the short name in the paragraph (`Terra is a…`). This is opposite to the other skins.

cahrens replied:
Maybe ping UX again on whether they really want Terra to be the opposite of the others. If they do, that's fine.

cahrens commented on src/components/common.js:
```diff
@@ -523,15 +523,27 @@ export const HeroWrapper = ({ showMenu = true, bigSubhead = false, children }) =
         heavyWrapper('access data'), ', ', heavyWrapper('run analysis tools'), ', ',
-        span({ style: { whiteSpace: 'nowrap' } }, ['and', heavyWrapper(' collaborate'), '.'])
+        span({ style: { whiteSpace: 'nowrap' } }, ['and', heavyWrapper(' collaborate'), '. ']),
+        showDocLink ?
```
`showDocLink` is always false, so this code is not used.

rtitle replied:
Oops! I was intending this to be set to `true` from `SignInPage` but not `LandingPage` (since the latter already includes the links). Will fix.

cahrens commented on src/components/AuthStoreSetter.js:
```diff
@@ -0,0 +1,28 @@
+  const auth = useAuth()
+  useOnMount(() => authStore.update(_.set(['authContext'], auth)))
+
+  useEffect(() => {
```
Do these need to be separate useEffect blocks? They are all gated on `auth` changing.

rtitle replied:
I did it this way because each of the callback registrations returns a cleanup function which I wanted to hook into `useEffect` cleanup. There may be a way to do it in 1 useEffect block though. (And need to investigate the console errors you noted above too.)

cahrens replied:
Ah, I see.

cahrens replied:
The recommendation on this is to combine into a single `useEffect`. Along these lines, but this hasn't been tested yet.

````
  useEffect(() => {
    const cleanupFns = [
      auth.events.addAccessTokenExpiring(reloadAuthToken),
      auth.events.addUserLoaded(processUser),
      auth.events.addUserUnloaded(processUser)
    ]
    return () => _.forEach(fn => fn(), cleanupFns)
  }, [auth])

rtitle replied:
Nice, will do.

cahrens commented on src/components/AuthStoreSetter.js:
```diff
@@ -0,0 +1,28 @@
+  }, [auth])
+
+  useEffect(() => {
+    return auth.events.addUserLoaded(processUser)
```
Why are you calling both `addUserLoaded` and `addUserUnloaded`?

rtitle replied:
These are installing listeners on user sign-in and sign-out events. We need to handle both events to update our auth store.

cahrens commented on src/components/AuthStoreSetter.js:
```diff
@@ -0,0 +1,28 @@
+    return auth.events.addUserUnloaded(processUser)
+  }, [auth])
+
+  return null
```
Why do you need an explicit return of `null`?

rtitle replied:
This came from Pete's suggestion. :) This component does some side effects but doesn't render anything. Should it return something other than `null`?

cahrens replied:
I don't know-- @petesantos can you explain why there needs to be a return of null? I personally would love a comment in the code that explains this.

cahrens replied:
OK, so this is because we aren't actually rendering anything in this bastardization of a component. Could you add a comment in the code to make this clear?

cahrens commented on src/components/TopBar.js:
```diff
@@ -187,16 +186,14 @@ const TopBar = ({ showMenu = true, title, href, children }) => {
-                }, ['SIGN IN']) :
-                h(SignInButton)
+              h(Clickable, {
+                ...(isDatastage() || isBioDataCatalyst()) ? { href: Nav.getLink('workspaces') } : { onClick: signIn },
```
I don't really understand what the experience is supposed to be for BioDataCatalyist. The text says "LOG IN", but you go to the workspaces page instead of directly logging in. And there we see "LOGIN".

![image](https://user-images.githubusercontent.com/484484/167467637-99a09556-cf8b-4622-9761-8378668b9bd3.png)

rtitle replied:
I'm not sure either honestly but that behavior predated my change, so I kept it.

cahrens replied:
Can you make "LOGIN"/"LOG IN" consistent? Before, the behavior was "SIGN IN" first, then "LOGIN".

cahrens replied:
Asking around, the folklore is that it is necessary to redirect to the page because BioData Catalyst requires to have the disclaimer text visible when you choose to log in (and Datastage is actually just an older name of BioData Catalyst).

cahrens commented:
I think you want to catch the error when the user closes the Google auth window (without choosing an account).

PopupWindow.ts:41 Uncaught (in promise) Error: Popup closed by user
    at PopupWindow.ts:41:1

cahrens commented on src/components/common.js:
```diff
@@ -523,15 +523,27 @@ export const HeroWrapper = ({ showMenu = true, bigSubhead = false, children }) =
+        span({ style: { whiteSpace: 'nowrap' } }, ['and', heavyWrapper(' collaborate'), '. ']),
+        showDocLink ?
+          h(Link, {
+            'aria-label': 'Learn more about Terra',
```
Don't need `aria-label` here, the link text will be read.

cahrens commented on src/libs/auth.js:
```diff
@@ -52,8 +87,7 @@ const becameRegistered = (oldState, state) => {
-    const options = new window.gapi.auth2.SigninOptionsBuilder({ scope: 'https://www.googleapis.com/auth/cloud-billing' })
-    await getAuthInstance().currentUser.get().grant(options)
+    await signIn(true)
     // Wait 250ms before continuing to avoid errors due to delays in applying the new scope grant
```
I wonder if this delay is still needed given the different API being used.

rtitle replied:
it is not!

cahrens commented:
@rtitle I think that it is necessary to talk to UX about what we do on the billing projects page now with regard to auth having to be elevated. It's quite confusing now that we show "No Access" both when the account does not have access, and when there is no account associated. And the only way to elevate your permissions is to choose "Change Account" or "Remove Account". So if you just go to Billing and want to _see_ what account is linked, you have to choose to change the account… which is not at all obvious. These would be my recommendations--

1) Add a menu option that allows you to elevate the permissions (I'm not sure what one would call this).
2) Differentiate the text displayed for No Access vs. No Account.

rtitle commented:
@cahrens thanks I'll check with UX -- agree that "I don't have access" and "I haven't logged in yet" are a bit overloaded.

I also was just experimenting and found that I can grant the additional scopes with `signInSilent`, which doesn't generate the pop-up. So it feels a bit closer to how it currently works. I'm asking around to see if this is okay though..

cahrens commented:
@rtitle If you can get back to the state where the account elevation is not exposed to the user (so there is no degradation in behavior), then I think that the issue of "No Access/No Account" should be tackled separately-- I can create the ticket for it. I don't want to drag this PR on longer than necessary or make you do extra work!

rtitle committed:
Latest PR feedback

rtitle commented:
@cahrens I addressed the latest feedback in https://github.com/DataBiosphere/terra-ui/pull/2949/commits/4db0942c6245927982a44c0dbe04c851af2df006.

I still get the console error -- I think it has something to do with components that do:
```js
const auth = useStore(authStore)
```

However I don't think it's specific to this branch. I ran the `dev` branch locally and see the same error from `SignInButton.js` and `TopBar.js`. And I do _not_ see it in https://bvdp-saturn-dev.appspot.com/ or https://pr-1-dot-bvdp-saturn-dev.appspot.com/ (which is running this code). So maybe it has something to do with running locally?

cahrens commented:
Since the button now says "Login" instead of "Sign In", can you change the text above the button in DashboardPublic?

<img width="542" alt="image" src="https://user-images.githubusercontent.com/484484/167528930-dde2f3b8-e267-4d99-9ed0-b55acbb431e1.png">

rtitle committed:
Fix text in DashboardPublic

rtitle commented:
Done

cahrens commented:
We need to decide and be consistent-- are we going with "Login" or "Log in"?

![image](https://user-images.githubusercontent.com/484484/167632798-f35deba2-eee0-4026-9355-f885e1cd9d7a.png)
![image](https://user-images.githubusercontent.com/484484/167632892-61349d52-610e-4574-bd75-ee74ee138eb5.png)
![image](https://user-images.githubusercontent.com/484484/167632974-9829d575-8cb8-4d65-979b-d32d41e38b66.png)

cahrens commented:
@rtitle the billing access isn't working on the PR site. When I first go to the billing page, I'm still not seeing the billing accounts associated projects I own (it says "No Access"). And then when I click "Change Billing Account", I just get a spinner and nothing happens.

I really think that if we aren't going to be able to show the billing account without the user doing _some action_ that results in the auth escalation, we have to work with UX on how to present this. I think there would need to be a clear action that says "show me the billing account", separate from "Change or Delete the Billing Account." I'm sorry if I was unclear before-- I thought you were indicating that you found a way to get exactly the same behavior as before, and in that case I would be fine with improving the experience separately. But if there is a degradation, I do think we need to pull in UX now.

rtitle commented:
Thanks for raising this @cahrens. I pinged Lou for both of these, will circle back.

Also not sure why it's not working on the PR site, the sign-in is being blocked by `Refused to display 'https://pr-1-dot-bvdp-saturn-dev.appspot.com/' in a frame because it set 'X-Frame-Options' to 'deny'.`. Looking more..

kyuksel commented:
@rtitle It looks like the [removal of the usage of the library apis.google.com/js/platform.js](https://github.com/DataBiosphere/terra-ui/pull/2949/files#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdL9) takes care of the maintenance ticket [[Terra UI] Implement alternative to auth library soon to be deprecated](https://broadworkbench.atlassian.net/browse/TUIM-1) as well  🎉  I verified that I don't see the warning on the console with your branch. I'm taking the liberty to prefix the PR title with the ticket and assign it to you. Please let me know if I'm missing something.

rtitle committed:
Per Lou, should be 'LOG IN'

rtitle committed:
STill need to run by Lou but I think this might be a nice solution for the billing scopes

rtitle committed:
Does this fix the test?

rtitle committed:
Fix silent auth refresh in PR sites -- PENDING APPSEC APPROVAL

rtitle commented:
@cahrens OK I have a little more clarity on the billing scope stuff, and a good solution, I think.

The extra scope only needs to be granted once per user/app. This is why when you log in as hermione on dev you can see billing information immediately -- she has already granted the scope so doesn't need to re-log in. If you create a new Google user and navigate to the billing screens then it generates a popup to grant the additional scopes (which is actually blocked on non-prod because Google hasn't approved it).

With that in mind I think this PR now handles both cases:

1. The user has already granted the scope for the app -- we silently add it on billing page load, so the user sees billing info immediately with no extra interaction.

2. The user has not granted the scope -- the user sees a `View billing account` link which when clicked will generate the pop-up to login and grant the scopes.

Video of 1: https://user-images.githubusercontent.com/5368863/167855976-a3f00bab-a049-4c66-afc5-6becdca64153.mov

Video of 2 (note the actual grant is blocked by Google because it's a local env): https://user-images.githubusercontent.com/5368863/167856085-d4d5fcad-f515-42b6-a262-57a964c2c031.mov

Lou has 👍  this UX, and I've coordinated with Denis and Muscles too. The good news is this approach will work with B2C as well, which will use a separate policy for the sensitive scopes.

By the way -- thanks for all your patience and thoroughness on this PR review! This stuff is tricky, and I really appreciate it.

rtitle commented on app.yaml:
```diff
@@ -13,7 +13,7 @@ handlers:
   secure: always
   http_headers:
-    X-Frame-Options: "DENY"
+    X-Frame-Options: "SAMEORIGIN"
```
@dinvlad I want to get your thoughts on this. The new auth library renders an iframe to do silent token refreshes, which was being blocked on AppEngine. I confirmed this change fixes it. The original PR to add `X-Frame-Options: DENY` is here: https://github.com/DataBiosphere/terra-ui/pull/1935.

Does AppSec have any concerns about relaxing this to `SAMEORIGIN`?

rtitle replied:
@sarahgibs approved this in slack https://broadinstitute.slack.com/archives/CADU7L0SZ/p1652276452198059

cahrens commented on src/libs/auth.js:
```diff
@@ -52,13 +98,14 @@ const becameRegistered = (oldState, state) => {
   }
 }
 
+const becameRegistered = (oldState, state) => {
```
This just moved locations in the file?

rtitle replied:
Ah yeah was just bunching the scope-related methods together

cahrens commented on src/pages/billing/Project.js:
```diff
@@ -651,6 +651,11 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
           ])])
         ])
       ]),
+      Auth.hasBillingScope() || div({ style: { color: colors.dark(), fontSize: 14, display: 'flex', alignItems: 'center', marginTop: '0.5rem', marginLeft: '1rem' } }, [
```
@rtitle I like this solution very much! I watched the videos and tried it out, and it's so much clearer.

As a nit, I find the || above somewhat difficult to read and think think that `!Auth.hasBillingScope() &&` is a clearer way to express that the HTML is only rendered if `hasBillingScope` returns false.

cahrens commented on integration-tests/tests/billing-projects.js:
```diff
@@ -1,12 +1,15 @@
   return {
-    visit: async () => await testPage.goto(`${testUrl}/#billing`),
+    visit: async () => {
+      await testPage.goto(`${testUrl}/#billing`)
```
Could you change this to `noSpinnersAfter`? It is less likely to be flaky.

rtitle replied:
I tried this but put it back with a comment.. I don't think `noSpinnersAfter` works when the action is `page.goto(…)` because it tries to verify the spinner appears and disappears on a single page (and in this case the page is changing).

cahrens replied:
OK, thanks!

cahrens replied:
It's actually interesting because there is a similar conversation going on here-- https://github.com/DataBiosphere/terra-ui/pull/3021/files#r870656864. My thought with `noSpinnersAfter` was to avoid this sort of problem, but I understand why it doesn't work. I can keep an eye on the test though and try to change to the promise approach Alex is suggesting if we see failures.

cahrens commented on src/components/SignInButton.js:
```diff
@@ -1,44 +1,21 @@
-  })
+    onClick: signIn,
+    style: { marginTop: '0.875rem', width: '9.4rem', height: '3.2rem', fontSize: '1rem' }
+  }, ['log in'])
```
Note: I created https://broadworkbench.atlassian.net/browse/SUP-676 for Terra Data Repository to also standardize on "Log in".

cahrens reviewed (approved):
Thank you for the attention to detail on this PR @rtitle! I don't need to re-review after the requested test code change.

@petesantos mentioned he might take another look at the PR, so it might be worth pinging him before you merge.

rtitle committed:
PR feedback

rtitle committed:
Put back test the way it was

petesantos commented on src/components/AuthStoreSetter.js:
```diff
@@ -0,0 +1,28 @@
+      auth.events.addUserLoaded(processUser),
+      auth.events.addUserUnloaded(processUser)
+    ]
+    return _.over(cleanupFns)
```
As written, this will not return a function to invoke upon cleanup, but the results of running all of the `cleanupFns`. I believe the cleanup will be invoked when as soon as this is run, not when `auth` changes.

Side note: TIL about `_.over`

```suggestion
    return () => _.over(cleanupFns)
```

rtitle replied:
@petesantos are you sure? I think `_.over([...])` itself returns a function, and doesn't invoke anything immediately. I did a quick test:
```
var f1 = () => console.log("f1")
var f2 = () => console.log("f2")

// prints nothing
var x = _.over([f1, f2])

// prints f1, f2
x()

// prints nothing
var y = () => _.over([f1, f2])

// prints nothing
y()

// prints f1, f2
y()()
```

But lmk if I'm misunderstanding soemthing.

rtitle replied:
(And I think all the `cleanupFns` are of type `() => void`)

petesantos replied:
You are right! I haven't used `over` before and when I looked at the doc I misread it

petesantos commented on src/components/AuthStoreSetter.js:
```diff
@@ -0,0 +1,28 @@
+      auth.events.addUserLoaded(processUser),
+      auth.events.addUserUnloaded(processUser)
+    ]
+    return _.over(cleanupFns)
```
As written, this will not return a function to invoke upon cleanup, but the results of running all of the `cleanupFns`. I believe the cleanup will be invoked when as soon as this is run, not when `auth` changes.

Side note: TIL about `_.over`

```suggestion
    return () => _.over(cleanupFns)
```
